File: array/6_union_intersection.c
#include <stdio.h>
#include <stdlib.h>

void show(int *arr, int size, char *msg)
{
    printf("\n%s: ", msg);
    for (int i = 0; i < size; i++)
        printf("%d ", arr[i]);
}

// Function for union.
void compute_union(int *arr1, int *arr2, int size_1, int size_2)
{
    // Max possible size for union size_1 + size_2.
    int *result = (int *)malloc(sizeof(int) * (size_2 + size_1));
    // arr1 index i.
    // arr2 index j.
    // result index index.
    int index, i, j;
    index = i = j = 0;

    // modified merge function from merge sort.
    while ((i < size_1) && (j < size_2))
    {
        if (arr1[i] == arr2[j])
        {
            result[index++] = arr1[i++];
            j++;
        }
        else if (arr1[i] < arr2[j])
            result[index++] = arr1[i++];
        else if (arr1[i] > arr2[j])
            result[index++] = arr2[j++];
    }

    while (i < size_1)
        result[index++] = arr1[i++];
    while (j < size_2)
        result[index++] = arr2[j++];

    show(result, index, "UNION");
}

void compute_intersection(int *arr1, int *arr2, int size_1, int size_2)
{
    // Max possible size for intersection min(size_1, size_2).
    int *result = (int *)malloc(sizeof(int) * (size_1 < size_2 ? size_1 : size_2));
    // arr1 index i.
    // arr2 index j.
    // result index index.
    int index, i, j;
    index = i = j = 0;

    // modified merge function from merge sort.
    while ((i < size_1) && (j < size_2))
    {
        if (arr1[i] == arr2[j])
        {
            result[index++] = arr1[i++];
            j++;
        }
        else if (arr1[i] < arr2[j])
            i++;
        else if (arr1[i] > arr2[j])
            j++;
    }

    show(result, index, "INTERSECTION");
}

void main()
{
    int arr1[] = {1, 3, 4, 5, 7};
    int arr2[] = {2, 3, 5, 6};
    // int arr1[] = {2, 5, 6};
    // int arr2[] = {4, 6, 8, 10};
    int size_1 = sizeof(arr1) / sizeof(int);
    int size_2 = sizeof(arr2) / sizeof(int);
    compute_union(arr1, arr2, size_1, size_2);
    compute_intersection(arr1, arr2, size_1, size_2);
    printf("\n");
}

File: python3/array/6_union_intersection.py
# using modified merge function
def compute_union(arr1, arr2):
    union = []
    index1 = 0
    index2 = 0

    while (index1 < len(arr1)) and (index2 < len(arr2)):
        if arr1[index1] < arr2[index2]:
            union.append(arr1[index1])
            index1 += 1
        elif arr1[index1] > arr2[index2]:
            union.append(arr2[index2])
            index2 += 1
        else:
            union.append(arr2[index2])
            index1 += 1
            index2 += 1

    while index1 < len(arr1):
        union.append(arr1[index1])
        index1 += 1
    while index2 < len(arr2):
        union.append(arr2[index2])
        index2 += 1

    return union


# using modified merge function
def compute_intersection(arr1, arr2):
    intersection = []
    index1 = 0
    index2 = 0

    while (index1 < len(arr1)) and (index2 < len(arr2)):
        if arr1[index1] < arr2[index2]:
            index1 += 1
        elif arr1[index1] > arr2[index2]:
            index2 += 1
        else:
            intersection.append(arr2[index2])
            index1 += 1
            index2 += 1

    return intersection


if __name__ == "__main__":
    arr1 = [1, 3, 4, 5, 7]
    arr2 = [2, 3, 5, 6]
    # arr1=[2, 5, 6]
    # arr2=[4, 6, 8, 10]
    union = compute_union(arr1, arr2)
    print('union : ', union)
    intersection = compute_intersection(arr1, arr2)
    print('intersection : ', intersection)
